physics/cires_vert_orodis_v1.F90

```diff
@@ -0,0 +1,1026 @@
+           endif
+         else
+           return                          ! wave absorption uf2 <= 0.
+         endif
+!
+! upward solver for single "mode" with tauz(levs+1) =0. at the top 
+! 
+         kds   =   0.1*  kvg(krefj)   ! eddy wave diffusion from the previous layer
+         kcrit =   levs
+         do k= krefj+1, levs  
+!
+! 2D-wave propagation along reference-wind direction
+!           udir = 0 critical wind for coro =0
+!           cdop = -uref .... upwind waves travel against MF
+!
+           udir =  ui(k)*xn +vi(k)*yn
+           uf2  =  udir*udir - c2f2
+        
+    
+           if (uf2 < dw2min .or. udir <= 0.0) then
+             kcrit =K
+             tauz(kcrit:levs) = 0. 
+             exit                    ! vert-level loop
+           endif
+!
+! wave-based solution
+!
+           mkz2 = bn2i(k)/uf2 -kx2
+           if (mkz2 > 0) then 
+             mkzm = sqrt(mkz2) 
+!
+! do dissipative flux vs saturation: kvg, ktg, krad, kion 
+!
+             kzw3 = mkzm*mkz2
+!
+             keff_m =  kvg(k)*mkz2 + kion(k)
+!            keff_t = kturb(k)*iPr_turb + kmol(k)*iPr_mol
+             keff_t = ktg(k)*mkz2  + krad(k)
+!       
+!
+             uf2p    = uf2 + 2.0*c2f2
+             betadis = uf2/uf2p
+             betaM   = 1.0 / (1.0+betadis)       ! if c2f2 = 0. betaM = betaT  =0.5 ekw = epw
+             betaT   = 1.0- BetaM
+
+!
+!imaginary frequencies of momentum and heat with "kds at (k-1) level"
+!
+             wfiM    = kds*mkz2  + keff_m
+             wfiT    = kds*mkz2  + keff_t
+!
+             cdfm    = sqrt(uf2)*kxs
+             cdft    = abs(udir)*kxs
+             wfdM    = wfiM/cdfm  *BetaM
+             wfdT    = wfiT/Cdft  *BetaT
+             mkzi    = 2.0*mkzm*(wfdM+wfdT)
+
+             fdis    = tauz(k-1)*exp(-mkzi*(zpi(k)-zpi(k-1)) )
+             tauz(k) = fdis
+             hsat2   = fcrit_sm2 * uf2 *bn2i(k)
+             fsat    = rhoi(k)* hsat2 * sqrt(uf2) * bvi(k)
+             if (fdis > fsat) then 
+               tauz(k)  = min(fsat, tauz(k-1))
+!=================================================================
+! two definitions for eddy mixing of MF:
+!  a) wave damping-Lindzen :  Ked ~ kx/(2H)*(u-c)^4/N^3
+!  b) heat-based turbulence: 4/3 Richardson Ked ~eps^1/3 *Lt^4/3 
+!=================================================================
+               kds     = rhp2*kxs*uf2*uf2/bn2i(k)/bvi(k)
+               kdis(k) = kds 
+             endif         
+           else
+             tauz(k:levs) = 0.     ! wave is reflected above
+             kds  = 0.
+           endif  
+         enddo
+
+         do k=krefj+1, kcrit
+           drtau(k) = rgrav*(tauz(k+1)-tauz(k))/dp(k)
+         enddo         
+!
+!
+      end  subroutine ugwp_oro_lsatdis
+!
+!
+     subroutine ugwp_tofd(im, levs, sigflt, elvmax, zpbl,  u, v, zmid, &
+                          utofd, vtofd, epstofd, krf_tofd)
+       use machine ,      only : kind_phys 
+       use ugwp_common_v1  , only :  rcpd2         
```

Comment:
Constants should come from the host model through the argument list, if possible.

Reply:
Done.